### DISCUSSION TITLE
Allow _FILE environment variables for Docker

### DISCUSSION THREAD

**Discussion by kevstev01** [directus/directus#5895] [Feature Requests]
In order to hide passwords and to use environment variables from files (i.e. docker-compose secrets), could the entrypoint be modified to include MySQL type environment file processing as per https://stackoverflow.com/questions/48094850/docker-stack-setting-environment-variable-from-secrets ? This would make password handling easier and prevent source code pollution for code repositories.

Cheers, Kevin

**Comment by rijkvanzanten** [comment]
Sounds like we'd just have to run this in the entrypoint right? `export $(egrep  -v '^#'  /run/secrets/* | xargs)`.

@WoLfulus any thoughts on this?

@kevstev01 wanna open a PR? 

@paescuj @t7tran worth adding this to the ongoing Docker improvements branch?

**Reply by WoLfulus** [reply]
This should be pretty straightforward to implement, but I'm unsure about putting it into the entrypoint vs the core codebase. It's very common to extend the docker image and change the entrypoint to deal with additional stuff and that would break if the user forgets to call or cannot call for some reason, the original entrypoint.

I can definitely see this as an extension that can come preinstalled into Docker images though, but we have to figure out the code that runs on imports first (actually there's a lot of stuff that opens up once we make this change, this is one of them).

**Reply by rijkvanzanten** [reply]
> vs the core codebase

How do you mean?

**Reply by WoLfulus** [reply]
> > vs the core codebase
> 
> How do you mean?

supporting _FILE suffixes on env.ts itself

**Reply by rijkvanzanten** [reply]
Not familiar with those. That just a MySQL Docker image thing?

**Reply by paescuj** [reply]
> @paescuj @t7tran worth adding this to the ongoing Docker improvements branch?

Would be better to address this in a separate pull request, IMO.

> Sounds like we'd just have to run this in the entrypoint right? `export $(egrep -v '^#' /run/secrets/* | xargs)`.

Hmm, I don't think this will work as expected...
Most official docker images have something like this in the entrypoint script:
```bash
# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
	local var="$1"
	local fileVar="${var}_FILE"
	local def="${2:-}"
	if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
		echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
		exit 1
	fi
	local val="$def"
	if [ "${!var:-}" ]; then
		val="${!var}"
	elif [ "${!fileVar:-}" ]; then
		val="$(< "${!fileVar}")"
	fi
	export "$var"="$val"
	unset "$fileVar"
}
```
And are then calling it like this:
```bash
file_env 'DB_PASSWORD'
```

But as @WoLfulus mentioned this could also be implemented in the application itself.

**Reply by rijkvanzanten** [reply]
Do these files (like `/run/secrets/rijk`) just contain a single string that we can use 1-1? Would we have a generic `_FILE` prefix for _every_ env var to dynamically load it from the file, or should it only be for certain ones?

**Reply by paescuj** [reply]
Yes, those files contain the secret value which can be used more or less 1-1.
Hmm, I was wondering the same. I can take a look at this tomorrow and maybe come up with a pull request...

**Reply by rijkvanzanten** [reply]
Sounds like we could go with a generic "If env var ends with `_FILE`, do a `await fs.readFile` to get the value" kinda thing

**Reply by paescuj** [reply]
Yep, was thinking about something like this as well.

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
